src/reflect/scala/reflect/internal/Positions.scala

```diff
@@ -139,17 +142,30 @@ trait Positions extends api.Positions { self: SymbolTable =>
             }
           }
         }
-        for (ct <- tree.children flatMap solidDescendants) validate(ct, tree)
+        for (ct <- childSolidDescendants) validate(ct, tree, trace)
       }
     }
 
     if (!isPastTyper)
-      validate(tree, tree)
+      validate(tree, tree, settings.Yposdebug && (settings.verbose || settings.Yrangepos))
   }
 
-  def solidDescendants(tree: Tree): List[Tree] =
-    if (tree.pos.isTransparent) tree.children flatMap solidDescendants
-    else List(tree)
+  final def solidDescendants(tree: Tree): List[Tree] =
+    if (tree.pos.isTransparent) {
+      val children = tree.children
+      if (children isEmpty) Nil else {
+        val result = new ListBuffer[Tree]
+        def solidDescendantsImpl(tree: Tree) : Unit = {
```

Comment:
I can't remember if this is inlined - if not you might be better to pass in the ```result``` buffer

Reply:
Captured parameters, like `result` here, are automatically added as extra method parameters by the compiler. No benefit in this case to do it manually.

Reply:
But given that the caller is:

```
tree.children flatMap solidDescendants
```

We would be better off with:

```scala
val solidDescendents = new ArrayBuffer[Tree]
tree.children.foreach(child => gatherSolidDescendants(child, solidDescendents)
...
```

Reply:
thanks - added a update to address this in https://github.com/scala/scala/pull/6696/commits/bc1b38772e273c917d10939bbb35296842772b99

Reply:
Thanks for the clarification @retronym